OPTM-24 New Development Section

firstdayofjune opened the pull request:
- described the workflow, CI pipeline and directory structure
- added tips & tricks (flutter format file watcher)

OPTM-24 #close

Commit:
New Development Section

- described the workflow, CI pipeline and directory structure
- added tips & tricks (flutter format file watcher)

OPTM-24 #start-review

guisedlythe commented:
I propose some minor changes/fixes:

````diff
diff --git a/README.md b/README.md
index 90261fc..dd25440 100644
--- a/README.md
+++ b/README.md
@@ -1,12 +1,12 @@
 # P2P Task Manager
 
-A privacy-aware task manager. No account needed, all your tasks get synced using the local network only. 
+A privacy-aware task manager. No account needed, all your tasks get synced using the local network only.
 
 ## Getting Started
 
-Follow the instruction on https://flutter.dev/docs/get-started/install
+Follow the instruction on <https://flutter.dev/docs/get-started/install>
 
-Follow the instructions on https://flutter.dev/docs/get-started/editor
+Follow the instructions on <https://flutter.dev/docs/get-started/editor>
 
 Run `flutter pub get` in the terminal to load all dependencies
 
@@ -22,14 +22,14 @@ A few resources to get you started:
 - [Cookbook: Useful Flutter samples](https://flutter.dev/docs/cookbook)
 - [Online Documentation](https://flutter.dev/docs)
 
-# Development
+## Development
 
 Read this section, if you want to contribute to the project.
 
 ## Workflow
 
 1. Select a JIRA issue to work on.
-2. Create a feature branch for the issue, named _OPTM-XY_Meaninful-Branch-Name_)
+2. Create a feature branch for the issue, named _OPTM-XY_Meaningful-Branch-Name_)
 3. Write your code and add tests for it
 4. Format your files (`flutter run format <filenames>` or `dart run format <filenames>`)
 5. Run all tests (`flutter test`) and make sure they complete without errors or warnings
@@ -44,6 +44,7 @@ Every file gets checked for proper formatting. Also, all tests get executed.
 You can find all the commands run on CI in `.github/workflows/main.yaml`.
 
 ## Directory Structure
+
 ```
 root
 // Contains the application code
@@ -82,6 +83,6 @@ In Visual Studio Code:
 > Install the Flutter extension (see [Editor setup](https://flutter.dev/docs/get-started/editor)) to get automatic formatting of code in VS Code.
 >
 > To automatically format the code in the current source code window, right-click in the code window and
-select Format Document. You can add a keyboard shortcut to this VS Code Preferences.
+> select Format Document. You can add a keyboard shortcut to this VS Code Preferences.
 >
 > To automatically format code whenever you save a file, set the _editor.formatOnSave_ setting to true.
````

JacobRoe reviewed (changes requested):
Beside the proposed changes by Konstantin, you might, in case you incorporate Konstantin's suggestions, want to adjust the child headers for the `Development` sections too.

I'm also not quite sure why we should use less then and greater than symbols for denoting a link. Typically, it gets rendered in blue anyway. Opinions? (Referring to Konstantin suggested changes.)

guisedlythe commented:
> I'm also not quite sure why we should use less then and greater than symbols for denoting a link. Typically, it gets rendered in blue anyway. Opinions? (Referring to Konstantin suggested changes.)

https://github.com/DavidAnson/markdownlint/blob/v0.23.1/doc/Rules.md#md034

JacobRoe commented:
> > I'm also not quite sure why we should use less then and greater than symbols for denoting a link. Typically, it gets rendered in blue anyway. Opinions? (Referring to Konstantin suggested changes.)
> 
> https://github.com/DavidAnson/markdownlint/blob/v0.23.1/doc/Rules.md#md034

Thank you. It also states the rationale at the end: "Without angle brackets, the URL isn't converted into a link by many markdown parsers." To be honest, it never happened to me that it didn't get correctly rendered (and I use markdown extensively). But at least now I see why it can make sense to add them. So, no objections anymore from my side.

Commit:
Merge remote-tracking branch 'origin/main' into OPTM-24_Dev-Section-in-Readme

# Conflicts:
#	README.md

Commit:
Instructions for git submodule and metrics

- documented git submodule in setup
- explained metrics in CI pipeline section
- explained usage of metrics

OPTM-24 #done

firstdayofjune commented:
I included the proposed changes and updated the README to reflect the new metrics in CI.

guisedlythe commented on README.md:
```diff
@@ -1,49 +1,57 @@
+With each Pull Request the following happens:
 
-You can find all the commands run on CI in `.github/workflows/main.yaml`.
+- Files are formatted using `flutter format`
```
Maybe a bit misleading: the CI checks if the files you committed are properly formatted, it doesn't actually format the files files for you.

guisedlythe commented on README.md:
```diff
@@ -1,49 +1,57 @@
 
-Follow the instructions on https://flutter.dev/docs/get-started/editor
+Clone this repository and open a terminal window in the root directory:
+- Run `git submodule init && git submodule update` (this project uses the lww_crdt library as a submodule)
```
Insert empty line before list: https://github.com/DavidAnson/markdownlint/blob/v0.23.1/doc/Rules.md#md032

guisedlythe commented on README.md:
````diff
@@ -1,49 +1,57 @@
+Find all the commands run on CI in `.github/workflows/main.yaml`.
+
+### Directory Structure
 ```
````
Insert empty line after heading: https://github.com/DavidAnson/markdownlint/blob/v0.23.1/doc/Rules.md#md022

guisedlythe commented on README.md:
```diff
@@ -72,16 +80,28 @@ You can create a file watcher to automatically format your files on save.
+
+Run the following command to get information about the code metrics
+
+    flutter pub run dart_code_metrics:metrics lib
```
Use consistent block style: https://github.com/DavidAnson/markdownlint/blob/v0.23.1/doc/Rules.md#md046
E.g.:
````
```sh
flutter pub run dart_code_metrics:metrics lib
```
````

guisedlythe commented on README.md:
```diff
@@ -1,49 +1,57 @@
 5. Run all tests (`flutter test`) and make sure they complete without errors or warnings
-6. Write a descriptive commit message and commit your changes (use the [JIRA Smart Commit syntax](https://support.atlassian.com/jira-software-cloud/docs/process-issues-with-smart-commits/) to reference the issue)
-7. Push your changes and create a pull request into main when you are done.
+6. Run code metrics (`flutter pub run dart_code_metrics:metrics lib`) and fix warnings related to your changes
```
Maybe just add a reference to the 'Run code metrics locally section'.
E.g.:
```md
[Run code metrics](#run-code-metrics-locally)
```

guisedlythe commented on README.md:
```diff
@@ -72,16 +80,28 @@ You can create a file watcher to automatically format your files on save.
+
+## Run code metrics locally
+
+Run the following command to get information about the code metrics
```
Maybe add that the following command runs the metrics only in the `lib` folder, use `flutter pub run dart_code_metrics:metrics lib test` to run it in the `test` folder as well.

firstdayofjune committed:
Updated readme according to review

- added missing blank lines
- added link
- explained CI more precise
- added example for code metrics on tests

guisedlythe committed:
OPTM-24 Fixed some empty line warnings

guisedlythe reviewed (approved):
Approved, except the minor changes that I already pushed to this branch.
